### PR TITLE
Dl 492

### PR DESCRIPTION
**Pull request by robert-webb-hmrc**
View Specs and helpers for the Confirmation, Existing Protections, and result pages.

### CODE REVIEW

File: test/views/pages/confirmation/confirmFPSpec.scala [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+import testHelpers.ViewSpecHelpers.confirmation.confirmFP
+import testHelpers.ViewSpecHelpers.CommonViewSpecHelper
+import testHelpers.ViewSpecHelpers.CommonMessages
+import views.html.pages.confirmation.{confirmFP => views}
+
+class confirmFPSpec extends CommonViewSpecHelper with confirmFP with CommonMessages {
+
+  "The FP 2016 page" should {
+
+    implicit lazy val view = views()
```
Comment (jameshforster) [inline test/views/pages/confirmation/confirmFPSpec.scala]: Please remove implicit tags if they are not required. Make sure this is the case for each other view spec as well.

File: test/views/pages/confirmation/confirmFPSpec.scala [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+        lazy val h1Tag = doc.select("H1")
+
+        s"have the page heading '$plaConfirmFP16PageHeading'" in {
+          h1Tag.text shouldBe {
+            plaConfirmFP16PageHeading
+          }
+        }
+
+        "have the heading-large class" in {
```
Comment (jameshforster) [inline test/views/pages/confirmation/confirmFPSpec.scala]: Classes that are purely visual (non-functional) do not need to be tested.

File: test/views/pages/confirmation/confirmFPSpec.scala [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+        s"have the link text $plaConfirmFP16HiddenParaLinkText ($plaBaseNewWindow)" in {
+          doc.body.select("details div p a").get(0).text shouldBe s"$plaConfirmFP16HiddenParaLinkText ($plaBaseNewWindow)"
+        }
+
+        "have the correct link destination" in {
+          accordionLink.attr("href") shouldBe "https://www.gov.uk/hmrc-internal-manuals/pensions-tax-manual/ptm093100"
+        }
+
+        "have the correct link ID" in {
```
Comment (jameshforster) [inline test/views/pages/confirmation/confirmFPSpec.scala]: ID tests are not useful as the ID here is not a required component (dynamic or not)

File: test/views/pages/confirmation/confirmFPSpec.scala [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+        lazy val declarationText = doc.body.select("div p strong").text
+        lazy val declarationStrong = doc.select("div p strong")
+        lazy val declarationElement = doc.select("div")
+
+        s"contain the text $plaConfirmFP16Declaration" in {
+          declarationText shouldBe plaConfirmFP16Declaration
+        }
+
+        "have the panel-indent class" in {
```
Comment (jameshforster) [inline test/views/pages/confirmation/confirmFPSpec.scala]: See previous comments regarding classes.

File: test/views/pages/confirmation/confirmFPSpec.scala [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+        s"have the button text '$plaBaseSubmitApplication'" in {
+          submitButton.text shouldBe plaBaseSubmitApplication
+        }
+
+        "be of type submit" in {
+          submitButton.attr("type") shouldBe "submit"
+        }
+
+        "have the class 'button'" in {
```
Comment (jameshforster) [inline test/views/pages/confirmation/confirmFPSpec.scala]: Ditto to above. Please check each test spec for similar issues.

File: test/views/pages/existingProtections/existingProtectionsSpec.scala [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+      s"contain the text '$plaExistingProtectionsPageHeading'" in {
+        h1Tag.text shouldBe plaExistingProtectionsPageHeading
+      }
+
+      "have the class" in {
+        h1Tag.hasClass("heading-xlarge") shouldBe true
+      }
+    }
+      /*If No previous protections exist ...*/
```
Comment (jameshforster) [inline test/views/pages/existingProtections/existingProtectionsSpec.scala]: Either uncomment and make sure this works, or delete this.

File: test/views/pages/result/manualCorrespondenceNeededSpec.scala [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+      s"have the heading text $plaMcNeededPageHeading" in {
+      h1Tag.text shouldBe plaMcNeededPageHeading
+      }
+
+      "have the heading-xlarge class" in {
+        h1Tag.hasClass("heading-xlarge") shouldBe true
+      }
+
+      "have the correct Id" in {
```
Comment (jameshforster) [inline test/views/pages/result/manualCorrespondenceNeededSpec.scala]: Not  useful ID

File: test/views/pages/result/manualCorrespondenceNeededSpec.scala [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    implicit lazy val view = views()
+    implicit lazy val doc = Jsoup.parse(view.body)
+
+    "have the correct title" in {
+      doc.title() shouldBe plaMcNeededTitle
+    }
+
+    "have the first heading which" should {
+
+      lazy val h1Tag = doc.select("div div header h1")
```
Comment (jameshforster) [inline test/views/pages/result/manualCorrespondenceNeededSpec.scala]: These selectors can be dramatically simplified on this spec (remove multiple divs, identify by unique components only)

File: test/views/pages/result/resultPrintSpec.scala [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+      "contain the following title message information" in {
+        tableHeading.get(0).text shouldBe plaPrintApplicationDate
+        tableHeading.get(2).text shouldBe plaPrintProtectionType
+        tableHeading.get(4).text shouldBe plaPrintPla
+        tableHeading.get(6).text shouldBe plaPrintProtectionNotificationNumber
+        tableHeading.get(8).text shouldBe plaPrintSchemeAdministratorReference
+      }
+
+      "contain the following id information" in {
```
Comment (jameshforster) [inline test/views/pages/result/resultPrintSpec.scala]: Note: This is a useful location for IDs so this would be permissable.

File: test/views/pages/result/resultSuccessInactiveSpec.scala [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+      "have the correct Id" in {
+        h1Tag.attr("id") shouldBe "resultOutcome"
+      }
+    }
+
+    "have a result code paragraph which" should {
+
+      "have the text" in {
+        doc.body.select("p").get(0).text  shouldBe "As you already have individual protection 2014 in place, fixed protection 2016 will only become active if you lose individual protection 2014."//Messages(s"$resultCode."+res.notificationId+".$infoNum")
```
Comment (jameshforster) [inline test/views/pages/result/resultSuccessInactiveSpec.scala]: Commented components need to be added to code or removed.

File: test/views/pages/result/resultSuccessInactiveSpec.scala [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+      "have the text" in {
+        doc.select("p").get(2).text shouldBe plaResultSuccessIPPensionSharing
+      }
+
+      "have the correct Id" in {
+        doc.select("p").get(2).attr("id") shouldBe "ipPensionSharing"
+      }
+    }
+
+      /*If Application Type = FP2016 ...*/
```
Comment (jameshforster) [inline test/views/pages/result/resultSuccessInactiveSpec.scala]: See earlier comments about commented code.

File: test/views/pages/result/resultSuccessSpec.scala [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+      "have the text" in {
+        doc.body.select("p").get(5).text shouldBe plaResultSuccessIPPensionSharing
+      }
+
+      "have the correct Id" in {
+        doc.body.select("p").get(5).attr("id") shouldBe "ipPensionSharing"
+      }
+    }
+
+      /*If Application Type = FP2016 ...*/
```
Comment (jameshforster) [inline test/views/pages/result/resultSuccessSpec.scala]: Commented code.

File: test/testHelpers/ViewSpecHelpers/confirmation/confirmFP.scala [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package testHelpers.ViewSpecHelpers.confirmation
+
+trait confirmFP {
```
Comment (jameshforster) [inline test/testHelpers/ViewSpecHelpers/confirmation/confirmFP.scala]: Please follow correct Trait naming scheme (capitalise the first letter)

File: test/testHelpers/ViewSpecHelpers/existingProtections/existingProtections.scala [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package testHelpers.ViewSpecHelpers.existingProtections
+
+trait existingProtections {
```
Comment (jameshforster) [inline test/testHelpers/ViewSpecHelpers/existingProtections/existingProtections.scala]: Capitalise first letter

File: test/testHelpers/ViewSpecHelpers/result/manualCorrespondenceNeeded.scala [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package testHelpers.ViewSpecHelpers.result
+
+trait manualCorrespondenceNeeded {
```
Comment (jameshforster) [inline test/testHelpers/ViewSpecHelpers/result/manualCorrespondenceNeeded.scala]: Again and repeat for all others.

File: test/views/pages/confirmation/confirmFPSpec.scala [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+
+import org.jsoup.Jsoup
+
+import play.api.i18n.Messages.Implicits._
+import testHelpers.ViewSpecHelpers.confirmation.confirmFP
+import testHelpers.ViewSpecHelpers.CommonViewSpecHelper
+import testHelpers.ViewSpecHelpers.CommonMessages
+import views.html.pages.confirmation.{confirmFP => views}
+
+class confirmFPSpec extends CommonViewSpecHelper with confirmFP with CommonMessages {
```
Comment (jameshforster) [inline test/views/pages/confirmation/confirmFPSpec.scala]: Classes must also have capitalised first letters. Also CommonMessages is imported by the ViewSpecHelper and is not needed as an extension here.

File: test/views/pages/existingProtections/existingProtectionsSpec.scala [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+import models.{ExistingProtectionDisplayModel, ExistingProtectionsDisplayModel}
+import org.jsoup.Jsoup
+import play.api.i18n.Messages
+import play.api.i18n.Messages.Implicits._
+import play.api.mvc.Call
+import testHelpers.ViewSpecHelpers.{CommonMessages, CommonViewSpecHelper}
+import testHelpers.ViewSpecHelpers.existingProtections.existingProtections
+import views.html.pages.existingProtections.{existingProtections => views}
+
+class existingProtectionsSpec extends CommonViewSpecHelper with existingProtections with CommonMessages {
```
Comment (jameshforster) [inline test/views/pages/existingProtections/existingProtectionsSpec.scala]: Existing protections has logic in the view based on the data passed in. In order to fully test this all combinations of data that require logic checking need to be tested as well (in this case, having both an active protection and empty active protection as well as a case for no other protections and for other protections.)

File: test/views/pages/result/resultPrintSpec.scala [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+
+import play.api.i18n.Messages.Implicits._
+import org.jsoup.Jsoup
+import testHelpers.ViewSpecHelpers.{CommonMessages, CommonViewSpecHelper}
+import testHelpers.ViewSpecHelpers.result.resultPrint
+import views.html.pages.result.{resultPrint => views}
+import models.PrintDisplayModel
+import play.api.i18n.Messages
+
+class resultPrintSpec extends CommonViewSpecHelper with resultPrint with CommonMessages {
```
Comment (jameshforster) [inline test/views/pages/result/resultPrintSpec.scala]: There is untested logic on this view (look for the map and the if statements)

File: test/views/pages/result/resultSuccessInactiveSpec.scala [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+import enums.ApplicationType
+import models.{ProtectionDetailsDisplayModel, RejectionDisplayModel, SuccessDisplayModel}
+import play.api.i18n.Messages.Implicits._
+import org.jsoup.Jsoup
+import play.api.i18n.Messages
+import testHelpers.ViewSpecHelpers.{CommonMessages, CommonViewSpecHelper}
+import testHelpers.ViewSpecHelpers.result.resultSuccessInactive
+import views.html.pages.result.{resultSuccessInactive => views}
+
+class resultSuccessInactiveSpec extends CommonViewSpecHelper with CommonMessages with resultSuccessInactive {
```
Comment (jameshforster) [inline test/views/pages/result/resultSuccessInactiveSpec.scala]: Missing logic tests: Look for the if and map statements.

File: test/views/pages/result/resultSuccessSpec.scala [hunk 20]
```diff
@@ -0,0 +1,204 @@
...
+import enums.ApplicationType
+import models.{ProtectionDetailsDisplayModel, SuccessDisplayModel}
+import testHelpers.ViewSpecHelpers.{CommonMessages, CommonViewSpecHelper}
+import testHelpers.ViewSpecHelpers.result.resultSuccess
+import views.html.pages.result.{resultSuccess => views}
+import org.jsoup.Jsoup
+import play.api.i18n.Messages.Implicits._
+
+
+class resultSuccessSpec extends CommonViewSpecHelper with CommonMessages with resultSuccess {
```
Comment (jameshforster) [inline test/views/pages/result/resultSuccessSpec.scala]: Missing logic tests: Look for the if and map statements.

File: app/views/pages/result/resultPrint.scala.html [hunk 21]
```diff
@@ -74,7 +74,7 @@ <h2>
 
             <table class="table">
                 <tbody>
-                @model.certificateDate.map { cDate =>
+                @model.certificateDate.map { cDate =>w
```
Comment (jameshforster) [inline app/views/pages/result/resultPrint.scala.html]: Is this for a reason or is it a typo?

File: test/views/pages/result/ResultSuccessSpec.scala [hunk 22]
```diff
@@ -96,6 +95,51 @@ class ResultSuccessSpec extends CommonViewSpecHelper with ResultSuccess {
...
+      }
+
+      "have a psa line(protection ref defined)" in {
+        doc2.select("ul li").get(3).attr("id") shouldBe "psaRef"
+        doc2.select("ul li").get(3).text shouldBe s"$plaResultSuccessPsaRef: PSA33456789"//@details.psaReference
+      }
+
+      "have an application date line(protection ref defined)" in {
+        doc2.select("ul li").get(4).attr("id") shouldBe "applicationDate"
+        doc2.select("ul li").get(4).text shouldBe s"$plaResultSuccessApplicationDate: 15/07/2015" //@details.applicationDate.get
```
Comment (jameshforster) [inline test/views/pages/result/ResultSuccessSpec.scala]: Comment introduced?
